crypto/evp_extra/evp_extra_test.cc

```diff
@@ -1695,3 +1699,512 @@ TEST_P(EVPRsaPssBadKeyTest, InvalidSaltLength) {
 
 INSTANTIATE_TEST_SUITE_P(All, EVPRsaPssBadKeyTest,
                          testing::ValuesIn(kBadPssKeyTestInputs));
+
+
+// START KEM TESTS
+
+struct KnownKEM {
+  const char name[20];
+  const int nid;
+  const size_t public_key_len;
+  const size_t secret_key_len;
+  const size_t ciphertext_len;
+  const size_t shared_secret_len;
+  const char *kat_filename;
+};
+
+static const struct KnownKEM kKEMs[] = {
+  {"Kyber512", NID_KYBER512, 800, 1632, 768, 32, "pq_kem_kat_tests_kyber512.txt"},
+};
+
+class PerKEMTest : public testing::TestWithParam<KnownKEM> {};
+
+INSTANTIATE_TEST_SUITE_P(All, PerKEMTest, testing::ValuesIn(kKEMs),
+                         [](const testing::TestParamInfo<KnownKEM> &params)
+                             -> std::string { return params.param.name; });
+
+TEST_P(PerKEMTest, KeyGeneration) {
+
+  // ---- 1. Test basic key generation flow ----
+  // Create context of KEM type.
+  bssl::UniquePtr<EVP_PKEY_CTX> ctx(EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, nullptr));
+  ASSERT_TRUE(ctx);
+
+  // Setup the context with specific KEM parameters.
+  int kem_nid = GetParam().nid;
+  ASSERT_TRUE(EVP_PKEY_CTX_kem_set_params(ctx.get(), kem_nid));
+
+  // Generate a key pair.
+  EVP_PKEY *raw = nullptr;
+  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
+  ASSERT_TRUE(EVP_PKEY_keygen(ctx.get(), &raw));
+  ASSERT_TRUE(raw);
+  bssl::UniquePtr<EVP_PKEY> pkey(raw);
+
+  // ---- 2. Test key generation with PKEY as a template ----
+  ctx.reset(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+  ASSERT_TRUE(ctx);
+
+  // Generate a key pair.
+  raw = nullptr;
+  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
+  ASSERT_TRUE(EVP_PKEY_keygen(ctx.get(), &raw));
+  ASSERT_TRUE(raw);
+  pkey.reset(raw);
```

Comment:
why reset? Should be unnecessary now at this point.

Reply:
added back. The `raw` pointer is a regular C pointer so we have to free in the end, otherwise the test leaks the allocated memory. Instead of freeing directly, I just added back the `reset` -- the `UniquePtr` class takes care of properly freeing the object.